src/directory_tree.cpp

```diff
@@ -0,0 +1,291 @@
+ * along with EasyRPG Player. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "directory_tree.h"
+#include "filefinder.h"
+#include "filefinder_rtp.h"
+#include "main_data.h"
+#include "output.h"
+#include "platform.h"
+#include "player.h"
+#include <lcf/reader_util.h>
+
+#ifdef EP_DEBUG_DIRECTORYTREE
+template <typename... Args>
+static void DebugLog(const char* fmt, Args&&... args) {
+	Output::Debug(fmt, std::forward<Args>(args)...);
+}
+#else
+template <typename... Args>
+static void DebugLog(const char*, Args&&...) {}
+#endif
+
+namespace {
+	std::string make_key(StringView n) {
+		return lcf::ReaderUtil::Normalize(n);
+	};
+}
+
+std::unique_ptr<DirectoryTree> DirectoryTree::Create() {
+	return std::make_unique<DirectoryTree>();
+}
+
+std::unique_ptr<DirectoryTree> DirectoryTree::Create(std::string path) {
+	// FIXME: Requires VFS handle passed in, currently hardcoded to FileFinder
+
+	if (!(FileFinder::Exists(path) && FileFinder::IsDirectory(path, true))) {
+		return std::unique_ptr<DirectoryTree>();
+	}
+	std::unique_ptr<DirectoryTree> tree = std::make_unique<DirectoryTree>();
+	tree->root = std::move(path);
+
+	DebugLog("Create: {}", tree->root);
+
+	return tree;
+}
+
+DirectoryTree::DirectoryListType* DirectoryTree::ListDirectory(StringView path) const {
+	std::vector<Entry> entries;
+	std::string fs_path = ToString(path);
+	std::string full_path = MakePath(path);
+
+	DebugLog("ListDirectory: {}", full_path);
+
+	auto dir_key = make_key(fs_path);
+
+	auto dir_it = dir_cache.find(dir_key);
+	if (dir_it != dir_cache.end()) {
+		// Already cached
+		DebugLog("ListDirectory Cache Hit: {}", dir_key);
+		auto file_it = fs_cache.find(dir_key);
+		assert(file_it != fs_cache.end());
+		return &file_it->second;
+	}
+
+	assert(fs_cache.find(dir_key) == fs_cache.end());
+
+	// FIXME: Skip this scan when the directory name matches the passed name
+
+	if (!fs_path.empty()) {
+		// Go up and determine the proper casing of the folder
+		std::string parent_dir, child_dir;
+		std::tie(parent_dir, child_dir) = FileFinder::GetPathAndFilename(fs_path);
+
+		auto* parent_tree = ListDirectory(parent_dir);
+		if (!parent_tree) {
+			return nullptr;
+		}
+
+		auto parent_key = make_key(parent_dir);
+		auto parent_it = dir_cache.find(parent_key);
+		assert(parent_it != dir_cache.end());
+
+		auto child_key = make_key(child_dir);
+		auto child_it = parent_tree->find(child_key);
+		if (child_it != parent_tree->end()) {
+			full_path = FileFinder::MakePath(MakePath(parent_it->second), child_it->second.name);
+			fs_path = FileFinder::MakePath(parent_it->second, child_it->second.name);
+		} else {
+			return nullptr;
+		}
+	}
+
+	Platform::Directory dir(full_path);
+	if (!dir) {
+		Output::Debug("Error opening dir {}: {}", full_path, ::strerror(errno));
+		return nullptr;
+	}
+
+	while (dir.Read()) {
+		const std::string name = dir.GetEntryName();
+		Platform::FileType type = dir.GetEntryType();
+
+		static bool has_fast_dir_stat = true;
```

Comment:
This dance with the static bool is confusing, is it really necessary? Why not just say 
```
if (type == Directory) {
  is_directory = true;
} else if (type == Unknown) {
  is_directory = FileFinder::IsDirectory(...);
}
```

Reply:
As this was copied over from FileFinder the code is likely 3 or more years old.
Your solution should work and looks better, thanks.
(The directory type in dirent is an extension and some filesystems reports Unknown here, will also add a comment)